pkg/storegateway/bucket.go

```diff
@@ -615,102 +621,308 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_Serie
+
+	sendBegin := time.Now()
+	if err := srv.SendMsg(msg); err != nil {
+		return status.Error(codes.Unknown, errors.Wrap(err, "send streaming series response").Error())
+	}
+	sendDuration += time.Since(sendBegin)
+
+	if seriesSet.Err() != nil {
+		return errors.Wrap(seriesSet.Err(), "expand series set")
+	}
+
+	return nil
+}
+
+func (s *BucketStore) sendSeriesChunks(
+	req *storepb.SeriesRequest,
+	srv storepb.Store_SeriesServer,
+	seriesSet storepb.SeriesSet,
+	stats *safeQueryStats,
+	iterationBegin time.Time,
+) (err error) {
+	var (
+		encodeDuration  = time.Duration(0)
+		sendDuration    = time.Duration(0)
+		seriesCount     int
+		chunksCount     int
+		streamingChunks = req.StreamingChunksBatchSize > 0
+	)
 
-		err = nil
+	if iterationBegin.Equal(time.Time{}) {
+		iterationBegin = time.Now()
+	}
+
+	// Once the iteration is done we will update the stats.
+	defer stats.update(func(stats *queryStats) {
+		stats.mergedSeriesCount += seriesCount
+		stats.mergedChunksCount += chunksCount
+
+		// The time spent iterating over the series set is the
+		// actual time spent fetching series and chunks, encoding and sending them to the client.
+		// We split the timings to have a better view over how time is spent.
+		stats.streamingSeriesFetchSeriesAndChunksDuration += stats.streamingSeriesWaitBatchLoadedDuration
+		stats.streamingSeriesEncodeResponseDuration += encodeDuration
+		stats.streamingSeriesSendResponseDuration += sendDuration
+		stats.streamingSeriesOtherDuration += time.Duration(util_math.Max(0, int64(time.Since(iterationBegin)-
+			stats.streamingSeriesFetchSeriesAndChunksDuration-encodeDuration-sendDuration)))
 	})
 
-	if err != nil {
-		return
+	var (
+		batchSizeBytes int
+		chunksBuffer   []*storepb.StreamingChunks
+	)
+	if streamingChunks {
+		chunksBuffer = make([]*storepb.StreamingChunks, req.StreamingChunksBatchSize)
+		for i := range chunksBuffer {
+			chunksBuffer[i] = &storepb.StreamingChunks{}
+		}
 	}
+	chunksBatch := &storepb.StreamingChunksBatch{
+		Series: chunksBuffer[:0],
+	}
+	for seriesSet.Next() {
+		// IMPORTANT: do not retain the memory returned by seriesSet.At() beyond this loop cycle
+		// because the subsequent call to seriesSet.Next() may release it.
+		lset, chks := seriesSet.At()
+		seriesCount++
+		var response *storepb.SeriesResponse
+		if streamingChunks {
+			// We only need to stream chunks here because the series labels have already
+			// been sent above.
+			chunksBatch.Series = chunksBatch.Series[:len(chunksBatch.Series)+1]
+			last := chunksBatch.Series[len(chunksBatch.Series)-1]
+			last.Chunks = append(last.Chunks[:0], chks...)
+			last.SeriesIndex = uint64(seriesCount - 1)
+			// Copy the chunk bytes to avoid race.
+			for i := range last.Chunks {
+				raw := last.Chunks[i].Raw
+				if raw == nil {
+					continue
+				}
+				newChk := storepb.Chunk{
+					Type: raw.Type,
+					Data: append(make([]byte, 0, len(raw.Data)), raw.Data...),
+				}
+				last.Chunks[i].Raw = &newChk
```

Comment:
This effectively doubles the allocations we spend for chunks.

I think we shouldn't be using a `SeriesSet` abstraction. Instead I think we should be using the existing iterators and their batch sizes. The existing `loadingSeriesChunksSetIterator` can load the chunks of X series. Currently `X` is set to `BucketStore.maxSeriesPerBatch`, we can set it to `req.StreamingChunksBatchSize` instead. This will also make sure that we load only a single batch per send instead of loading N batches to send 1 batch and also holding half-sent batches.

When one batch (aka one `seriesChunksSet`) is loaded, it is directly sent to the querier and it contains exactly `req.StreamingChunksBatchSize` of series (and their chunks).

This + the retention of the label sets while sending series makes me think that the `storepb.SeriesSet` abstraction is not working for streaming responses to the querier.

Sorry I'm only noticing this now, this should have probably been in the first round of review.

Reply:
I will see how to refactor this part. But sending exactly `req.StreamingChunksBatchSize` may not be a good idea and we need to have the size limit as well. I will try to make good use of a slice and optimise this.

Reply:
I have refactored to use the other iterator